docs/rfc/rfc-2.adoc

```diff
@@ -0,0 +1,300 @@
+Bob is added to the list of inactive members.
+
+If the majority does not agree that Bob failed to send the message,
+they will re-broadcast Bob's message to ensure that all members receive it.
+
+Whenever an operator is deemed to be inactive,
+the remaining participants start another key generation attempt
+with all inactive members removed.
+
+==== Dealing with misbehaviour
+
+The protocol in `tss-lib` is not attributable.
+When a participant is unable to proceed
+because they have received an invalid message from another participant,
+they will broadcast a special message declaring a _disqualification abort_
+and naming one or more participants who sent invalid messages.
+
+When a _disqualification abort message_ is broadcast by any member,
+group selection is aborted and no further attempts are made by the group.
+Instead, the group signs and submits an on-chain transaction
+announcing that key generation was aborted due to disqualification,
+and identifying the participants in the aborted attempt,
+the members who declared a disqualification abort
+and the members who were accused of sending invalid messages.
+All participants will also store their logs of the aborted attempt
+to facilitate later investigation.
+These logs must contain all messages sent and received by the participant,
+private keys the participant had established with all other participants,
+and all secret values the participant generated for the key generation attempt.
+
+Governance will then resolve the matter by investigating the stored logs
+and taking any action deemed appropriate as permitted by the contract.
+Governance may penalise any operators
+named in a disqualification abort transaction
+as either the accusing or accused parties.
+
+The `tss-lib` protocol is intended to be eventually replaced
+with a fully attributable key generation protocol.
+When the key generation protocol is replaced,
+the powers of governance to adjudicate misbehaviour off-chain
+and penalise operators deemed to have misbehaved
+are to be removed in favour of a fully objective resolution method.
+
+==== Successful key generation
+
+When an attempt at key generation finishes successfully,
+a transaction announcing this result is submitted on-chain.
+The transaction contains the public key from the successful DKG attempt,
+the list of participants in that successful attempt,
+and signatures from a majority of participants.
+
+The on-chain contract verifies
+that all listed participants are among the originally selected members,
+and that all provided signatures are valid signatures from different members.
+If the submission passes these checks,
+the newly created wallet is removed from the pending wallets
+and appended to the list of active wallets,
+its state is set to _Active_,
+and its activation time is recorded.
+
+If some participants were inactive in the key generation,
+it might be desirable to remove them from the sortition pool.
```

Comment:
Agreed. Right now there is no way to kick someone out and I think having this possibility is desirable. My only worry is that it probably needs to happen in a loop with up to N-C steps. If N-C is small enough, we should be fine. If not, we'll have to figure out something else.

Reply:
Adding the ability for a sortition pool owner to kick/ban operators (kick just removes and should be used for inactivity, ban permanently stops them joining) should be enough and pretty simple.

Reply:
Given that banning an operator and removing them costs around 75k gas, I think we have room for some looping. Agreed that we should ensure that we can afford it in any possible circumstance.